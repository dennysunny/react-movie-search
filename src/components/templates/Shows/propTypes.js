import { object, array, bool } from 'prop-types';

export const contextTypes = {
  router: object.isRequired,
};

export const propTypes = {
  shows: array.isRequired,
  genres: array.isRequired,
  loading: bool.isRequired,
  genresLoading: bool.isRequired,
  configLoading: bool.isRequired,
};

export const defaultProps = {
};
