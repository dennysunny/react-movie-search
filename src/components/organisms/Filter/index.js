/* eslint-disable camelcase */
import React, { useState, useEffect } from 'react';

import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import OutlinedInput from '@material-ui/core/OutlinedInput';
import { makeStyles } from '@material-ui/core/styles';

import { propTypes } from './propTypes';
import MyPager from '../../atoms/Pager';
import FilterGenres from '../../atoms/FilterGenres';

const useStyles = makeStyles(theme => ({
  'filter-wrapper': {
    '& form': {
      margin: `${theme.spacing(5)}px 0`,
    },
  },
}));

const objectToQueryStr = paramsObj => (
  Object.keys(paramsObj)
    .map(key => `${encodeURIComponent(key)}=${encodeURIComponent(paramsObj[key])}`)
    .join('&')
);

const initState = ({ params, release_year }) => ({
  page: parseInt(params.get('page'), 10) || 1,
  [release_year[0]]: params.get(release_year[0]) || new Date().getFullYear(),
  sort_by: params.get('sort_by') || 'popularity.desc',
  with_genres: params.get('with_genres') || '',
  'vote_average.gte': params.get('vote_average.gte') || 0,
});

const Filter = ({
  query,
  media,
  genres,
  children,
  queryUpdate
}) => {
  const classes = useStyles();
  const year = new Date().getFullYear();

  const params = new URLSearchParams(query);

  const release_year = media === 'tvs'
    ? ['first_air_date_year', 'first_air_date']
    : ['primary_release_year', 'primary_release_date'];

  const [state, setState] = useState(initState({ params, release_year }));

  useEffect(() => {
    const runQuery = () => {
      const strquery = objectToQueryStr(state);
      queryUpdate(strquery);
    };

    runQuery();
  }, [state, queryUpdate]);

  const sumYear = year - 1900;
  const yearOptions = Array(sumYear)
    .fill()
    .map((_, i) => (
      // eslint-disable-next-line react/no-array-index-key
      <MenuItem key={i} value={year - i}>
        {year - i}
      </MenuItem>
    ));

  const handleChange = (e) => {
    const val = e.target.value;
    const option = e.target.name;
    setState({ ...state, [option]: val, page: 1 });
  };

  const handleGenresChange = genres => setState({ ...state, with_genres: genres, page: 1 });

  const handlePageSelect = (page) => {
    if (page > 0) {
      setState({ ...state, page });
    }
  };

  const { sort_by, with_genres, page } = state;
  return (
    <div className={classes['filter-wrapper']}>
      <form>
        <FormControl
          variant="outlined"
          className="filter-element-wrapper"
        >
          <InputLabel htmlFor="shortby">
            Sort By
          </InputLabel>
          <Select
            id="shortby"
            placeholder="select"
            onChange={handleChange}
            value={sort_by}
            name="sort_by"
            input={<OutlinedInput labelWidth="53" id="shortby" />}
          >
            <MenuItem value="popularity.desc">Popularity Descending</MenuItem>
            <MenuItem value="popularity.asc">Popularity Ascending</MenuItem>
            <MenuItem value="vote_average.desc">Rating Descending</MenuItem>
            <MenuItem value="vote_average.asc">Rating Ascending</MenuItem>
            <MenuItem value={`${release_year[1]}.desc`}>
              Release Date Descending
            </MenuItem>
            <MenuItem value={`${release_year[1]}.asc`}>
              Release Date Ascending
            </MenuItem>
          </Select>
        </FormControl>
        <FormControl className="filter-element-wrapper" variant="outlined">
          <InputLabel>Year</InputLabel>
          <Select
            placeholder="select"
            onChange={handleChange}
            // eslint-disable-next-line react/destructuring-assignment
            value={state[release_year[0]]}
            name={release_year[0]}
            input={<OutlinedInput labelWidth="53" id="year" />}
          >
            <MenuItem value="null">None</MenuItem>
            {yearOptions}
          </Select>
        </FormControl>
        <FilterGenres
          genres={genres}
          onChange={handleGenresChange}
          value={with_genres}
        />
        <FormControl className="filter-element-wrapper" variant="outlined">
          <InputLabel labelWidth="200">Average vote</InputLabel>
          <Select
            name="vote_average.gte"
            placeholder="select"
            onChange={handleChange}
            // eslint-disable-next-line react/destructuring-assignment
            value={state['vote_average.gte']}
            input={<OutlinedInput labelWidth="200" id="vote" />}
            fullWidth
          >
            <MenuItem value="">Greater than...</MenuItem>
            <MenuItem value="9">9</MenuItem>
            <MenuItem value="8">8</MenuItem>
            <MenuItem value="7">7</MenuItem>
            <MenuItem value="6">6</MenuItem>
            <MenuItem value="5">5</MenuItem>
            <MenuItem value="4">4</MenuItem>
            <MenuItem value="3">3</MenuItem>
            <MenuItem value="2">2</MenuItem>
            <MenuItem value="1">1</MenuItem>
          </Select>
        </FormControl>
      </form>

      {children}

      <MyPager
        page={page}
        handlePageSelect={handlePageSelect}
      />
    </div>
  );
};

export default Filter;
