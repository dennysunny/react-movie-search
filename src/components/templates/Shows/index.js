import React, { useState } from 'react';

import { propTypes } from './propTypes';
import Filter from '../../organisms/Filter';
import Result from '../../molecules/Result';
import Spinner from '../../atoms/Spinner';
import PageTransition from '../../atoms/PageTransition';

const Shows = ({
  media,
  genres,
  loading,
  genresLoading,
  configLoading,
}) => {
  const [query, queryUpdate] = useState('');

  if (loading || genresLoading || configLoading) {
    return <Spinner />;
  }

  return (
    <PageTransition>
      <Filter
        genres={genres}
        queryUpdate={queryUpdate}
        media={media}
      >
        <Result
          query={query}
          media={media}
        />
      </Filter>
    </PageTransition>
  );
};

export default Shows;
