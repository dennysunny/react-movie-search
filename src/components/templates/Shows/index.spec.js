import React from 'react';
import { shallow } from 'enzyme';

import Filter from '../../organisms/Filter';
import Shows from '.';

const mockProps = {
  media: 'movies',
  genres: [{
    id: 1,
    name: 'genrename',
  }],
  query: '',
  config: {
    images: {
      secure_base_url: 'path/to/image',
      poster_sizes: ['1', '2', '3', '4'],
      backdrop_sizes: ['1', '2', '3', '4'],
    },
  },
  shows: [{}],
  loading: false,
  genresLoading: false,
  configLoading: false,
};

const context = { router: { history: ['/movies'] } };

describe('Templates/Shows', () => {
  it('should render as expected', () => {
    const wrapper = shallow(<Shows {...mockProps} />, { context });

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a spinner when loading', () => {
    const wrapper = shallow(<Shows {...mockProps} loading />, { context });
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle filter change', () => {
    const wrapper = shallow(<Shows {...mockProps} />, { context });
    wrapper.find(Filter).props().queryUpdate('page=2');

    expect(wrapper.context().router.history.includes('/movies?page=2')).toBe(true);
  });
});
