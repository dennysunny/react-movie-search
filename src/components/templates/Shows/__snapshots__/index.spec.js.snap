// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Templates/Shows should render a spinner when loading 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Shows
    config={
      Object {
        "images": Object {
          "backdrop_sizes": Array [
            "1",
            "2",
            "3",
            "4",
          ],
          "poster_sizes": Array [
            "1",
            "2",
            "3",
            "4",
          ],
          "secure_base_url": "path/to/image",
        },
      }
    }
    configLoading={false}
    genres={
      Array [
        Object {
          "id": 1,
          "name": "genrename",
        },
      ]
    }
    genresLoading={false}
    loading={true}
    media="movies"
    query=""
    resultMedia="movie"
    shows={
      Array [
        Object {},
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "router": Object {
        "history": Array [
          "/movies",
        ],
      },
    },
  },
}
`;

exports[`Templates/Shows should render as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Shows
    config={
      Object {
        "images": Object {
          "backdrop_sizes": Array [
            "1",
            "2",
            "3",
            "4",
          ],
          "poster_sizes": Array [
            "1",
            "2",
            "3",
            "4",
          ],
          "secure_base_url": "path/to/image",
        },
      }
    }
    configLoading={false}
    genres={
      Array [
        Object {
          "id": 1,
          "name": "genrename",
        },
      ]
    }
    genresLoading={false}
    loading={false}
    media="movies"
    query=""
    resultMedia="movie"
    shows={
      Array [
        Object {},
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Filter
        genres={
          Array [
            Object {
              "id": 1,
              "name": "genrename",
            },
          ]
        }
        media="movies"
        query=""
        queryUpdate={[Function]}
      >
        <Result
          config={
            Object {
              "images": Object {
                "backdrop_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "poster_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "secure_base_url": "path/to/image",
              },
            }
          }
          configLoading={false}
          genres={
            Array [
              Object {
                "id": 1,
                "name": "genrename",
              },
            ]
          }
          genresLoading={false}
          loading={false}
          media="movie"
          query=""
          resultMedia="movie"
          shows={
            Array [
              Object {},
            ]
          }
        />
      </Filter>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Result
          config={
            Object {
              "images": Object {
                "backdrop_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "poster_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "secure_base_url": "path/to/image",
              },
            }
          }
          configLoading={false}
          genres={
            Array [
              Object {
                "id": 1,
                "name": "genrename",
              },
            ]
          }
          genresLoading={false}
          loading={false}
          media="movie"
          query=""
          resultMedia="movie"
          shows={
            Array [
              Object {},
            ]
          }
        />,
        "genres": Array [
          Object {
            "id": 1,
            "name": "genrename",
          },
        ],
        "media": "movies",
        "query": "",
        "queryUpdate": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "config": Object {
            "images": Object {
              "backdrop_sizes": Array [
                "1",
                "2",
                "3",
                "4",
              ],
              "poster_sizes": Array [
                "1",
                "2",
                "3",
                "4",
              ],
              "secure_base_url": "path/to/image",
            },
          },
          "configLoading": false,
          "genres": Array [
            Object {
              "id": 1,
              "name": "genrename",
            },
          ],
          "genresLoading": false,
          "loading": false,
          "media": "movie",
          "query": "",
          "resultMedia": "movie",
          "shows": Array [
            Object {},
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Filter
          genres={
            Array [
              Object {
                "id": 1,
                "name": "genrename",
              },
            ]
          }
          media="movies"
          query=""
          queryUpdate={[Function]}
        >
          <Result
            config={
              Object {
                "images": Object {
                  "backdrop_sizes": Array [
                    "1",
                    "2",
                    "3",
                    "4",
                  ],
                  "poster_sizes": Array [
                    "1",
                    "2",
                    "3",
                    "4",
                  ],
                  "secure_base_url": "path/to/image",
                },
              }
            }
            configLoading={false}
            genres={
              Array [
                Object {
                  "id": 1,
                  "name": "genrename",
                },
              ]
            }
            genresLoading={false}
            loading={false}
            media="movie"
            query=""
            resultMedia="movie"
            shows={
              Array [
                Object {},
              ]
            }
          />
        </Filter>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Result
            config={
              Object {
                "images": Object {
                  "backdrop_sizes": Array [
                    "1",
                    "2",
                    "3",
                    "4",
                  ],
                  "poster_sizes": Array [
                    "1",
                    "2",
                    "3",
                    "4",
                  ],
                  "secure_base_url": "path/to/image",
                },
              }
            }
            configLoading={false}
            genres={
              Array [
                Object {
                  "id": 1,
                  "name": "genrename",
                },
              ]
            }
            genresLoading={false}
            loading={false}
            media="movie"
            query=""
            resultMedia="movie"
            shows={
              Array [
                Object {},
              ]
            }
          />,
          "genres": Array [
            Object {
              "id": 1,
              "name": "genrename",
            },
          ],
          "media": "movies",
          "query": "",
          "queryUpdate": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "config": Object {
              "images": Object {
                "backdrop_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "poster_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "secure_base_url": "path/to/image",
              },
            },
            "configLoading": false,
            "genres": Array [
              Object {
                "id": 1,
                "name": "genrename",
              },
            ],
            "genresLoading": false,
            "loading": false,
            "media": "movie",
            "query": "",
            "resultMedia": "movie",
            "shows": Array [
              Object {},
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "router": Object {
        "history": Array [
          "/movies",
        ],
      },
    },
  },
}
`;
